### ISSUE TITLE
USD RuntimeError: unidentifiable C++ exception compiling for UnrealEngine with clang

### ISSUE THREAD

**Issue by msearl001** [PixarAnimationStudios/USD#1033]
### Description of Issue
Hi I'm able to succesfully compile USD for UnrealEngine, but when importing to Python I get weird errors which I can't parse, because I don't get enough information about the error.

### Steps to Reproduce
1. Download UE4 from github and run Setup.sh
2. Goto Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build
3. Run GetUsd.sh
4. Open BuildUsd.sh and add at the beginning:
```
UE_ENGINE_DIR=`cd "../../../../../../.."; pwd`
COMPILER_PATH=$UE_ENGINE_DIR/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v15_clang-8.0.1-centos7/x86_64-unknown-linux-gnu
CMAKE=$UE_ENGINE_DIR//opt/build/UnrealEngine/Engine/Extras/ThirdPartyNotUE/CMake/bin

echo UE_ENGINE_DIR=$UE_ENGINE_DIR
echo COMPILER_PATH=$COMPILER_PATH

export PATH=$PYTHON_INSTALL_DIR/bin:$COMPILER_PATH/bin:$CMAKE/bin:$PATH
```
5. Run BuildUsd.sh

### System Information (OS, Hardware)
CentOS 7.6
Unreal Engine uses clang v15 with cmake

### Package Versions
v0.7.5
v19.01
v19.11

### Build Flags
```
	export CXXFLAGS="-std=c++11 -fPIC -nostdinc++ -I$UE_THIRD_PARTY_DIR/Linux/LibCxx/include -I$UE_THIRD_PARTY_DIR/Linux/LibCxx/include/c++/v1"
	CMAKE_LINKER_FLAGS="-stdlib=libc++ -L$UE_THIRD_PARTY_DIR/Linux/LibCxx/lib/Linux/x86_64-unknown-linux-gnu/ $UE_THIRD_PARTY_DIR/Linux/LibCxx/lib/Linux/x86_64-unknown-linux-gnu/libc++abi.a $COMPILER_PATH/lib64/libutil.so.1 $COMPILER_PATH/lib64/librt.so.1"
	cmake -DCMAKE_INSTALL_PREFIX=$USD_INSTALL_LOCATION  -DPXR_INSTALL_LOCATION=$USD_INSTALL_LOCATION -DPYTHON_INCLUDE_DIR=$PYTHON_INSTALL_DIR/include/python2.7 -DPYTHON_LIBRARY=$PYTHON_INSTALL_DIR/lib/libpython2.7.a -DCMAKE_CXX_FLAGS_RELEASE="$CXXFLAGS" -DCMAKE_EXE_LINKER_FLAGS="$CMAKE_LINKER_FLAGS" -DCMAKE_MODULE_LINKER_FLAGS="$CMAKE_LINKER_FLAGS" -DCMAKE_SHARED_LINKER_FLAGS="$CMAKE_LINKER_FLAGS" -DCMAKE_STATIC_LINKER_FLAGS="$CMAKE_LINKER_FLAGS" -DPXR_BUILD_IMAGING=FALSE -DPXR_BUILD_USD_IMAGING=FALSE -DPXR_BUILD_TESTS=FALSE -DBOOST_ROOT=../$BOOST_VER -DTBB_INCLUDE_DIRS=../tbb-$TBB_VER/include -DTBB_LIBRARIES=../tbb-$TBB_VER/build/localbuild_release -DTBB_tbb_LIBRARY=$TBB_ABS_ROOT/build/localbuild_release/libtbb.so --config Debug ../USD
	make install -j $MAX_JOBS
```

### Error for: from pxr import Vt:
```
Python 2.7.13 (default, Nov 13 2019, 12:42:33) 
[GCC 4.2.1 Compatible Clang 8.0.1 (tags/RELEASE_801/final)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pxr import Tf
>>> from pxr import Vt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Vt/__init__.py", line 31, in <module>
    import _vt
RuntimeError: unidentifiable C++ exception
```

### Error for: from pxr import Usd:
```
Python 2.7.13 (default, Nov 13 2019, 12:42:33) 
[GCC 4.2.1 Compatible Clang 8.0.1 (tags/RELEASE_801/final)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pxr import Usd
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Tf'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 85, in <module>
    from . import _tf
TypeError: No Python class registered for C++ class pxrInternal_v0_7__pxrReserved__::TfType
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Gf'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Gf/__init__.py", line 31, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Work'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Work/__init__.py", line 32, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Plug'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Plug/__init__.py", line 29, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Vt'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Vt/__init__.py", line 32, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Ar'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Ar/__init__.py", line 25, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Kind'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Kind/__init__.py", line 27, in <module>
    from pxr import Tf
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Tf/__init__.py", line 97, in <module>
    __SetErrorExceptionClass(ErrorException)
NameError: name '__SetErrorExceptionClass' is not defined
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Sdf'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Sdf/__init__.py", line 24, in <module>
    import _sdf
TypeError: No to_python (by-value) converter found for C++ type: pxrInternal_v0_7__pxrReserved__::TfToken
Warning: in Tf_PyLoadScriptModule at line 121 of /opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/USD/pxr/base/lib/tf/pyUtils.cpp -- Import failed for module 'pxr.Pcp'!
Traceback (most recent call last):
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Pcp/__init__.py", line 24, in <module>
    import _pcp
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/build/UnrealEngine/Engine/Plugins/Importers/USDImporter/Source/ThirdParty/Linux/build/InstalledUSD/lib/python/pxr/Usd/__init__.py", line 24, in <module>
    import _usd
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

I am aware that this is not your problem building USD for UnrealEngine, although I'm like to ask you to give me some hints on how to better debug the errors and what they could mean.
Maybe there are still some environment libraries picked up during compilation which cause this issue. Can you tell how to maybe enable some flags during build process to get more details about why it can't load in the lib files?

**Comment by jilliene** [comment]
Filed as internal issue #USD-5694

**Comment by msearl001** [comment]
Looks like it had issues with cmake 3.3. In you documentation it says cmake 2.8 - 3.6 for linux.
When using cmake 2.8 it worked in the end (after a long hours of retry and retry).

Thanks for filing this issue! I think there is some error reporting missing despite the fact of my cmake issue.

Cheers!

**Comment by c64kernal** [comment]
Hi @msearl001 -- I'm glad you figured out your issue. The only difference I can think of between 2.8 and 3.3 are how they find and pick up the version of Python installed on your system. Is it possible that you have multiple versions of python installed? In a standard cmake setup, you can pass in PYTHON_EXECUTABLE, PYTHON_INCLUDE_DIR, PYTHON_LIBRARY -- to specify exactly which python interpreter to use to build USD against.

Hope that helps!

**Comment by c64kernal** [comment]
(Please feel free to reopen or file a new issue if I've misunderstood the crux of the problem, of course!)

**Comment by msearl001** [comment]
> Hi @msearl001 -- I'm glad you figured out your issue. The only difference I can think of between 2.8 and 3.3 are how they find and pick up the version of Python installed on your system. Is it possible that you have multiple versions of python installed? In a standard cmake setup, you can pass in PYTHON_EXECUTABLE, PYTHON_INCLUDE_DIR, PYTHON_LIBRARY -- to specify exactly which python interpreter to use to build USD against.
> 
> Hope that helps!

Hi, thanks for your reply.
The PYTHON directives for cmake were given from EpicGames like this:
`-DPYTHON_INCLUDE_DIR=$PYTHON_INSTALL_DIR/include/python2.7 
-DPYTHON_LIBRARY=$PYTHON_INSTALL_DIR/lib/libpython2.7.a`

Maybe the PYTHON_EXECUTABLE is missing to make it work correctly?

But anyway with using cmake 2.8 it works. Thanks for your support :).

**Comment by sorenvdb** [comment]
Hi! I'm having the same issue, but changing the version of cmake to 2.8 has not fixed it. I'm using just cmake and not the python build script since all the dependencies needed to be installed manually in custom locations. 

@c64kernal  you said that multiple versions of python might cause this issue. Which versions/why exactly would that happen? I am using the same python version as the one specified in cmake, so could this still be the issue? 

Thanks for your help.

**Comment by msearl001** [comment]
When building for unreal make sure you use their build script, because they use their own libc, therefore everything has to be compiled against the libc version. If you don't do this you will get std library errors when loading up USD in Unreal.

Dependencies do not have to be installed manually, follow these steps to make it work:

1. Edit the `build/Versions.inc` and set the USD version to 19.01 (max 19.01, as any further doesn't implement the new API changes)
2. Edit the `build/GetUsd.sh` to clone the offical repository instead of the custom RCL repo
3. Run GetUsd.sh
4. Run BuildUsd.sh
5. You might need to add the compiler for UE4.24, as the build script is a little bit old and rusty. I've been also compiling Python with SSl (I had to download manually and built manually). Take out whatever code you need from this:

```
PYTHON_INSTALL_DIR=`pwd`/localpython

UE_ENGINE_DIR=`cd "../../../../../../.."; pwd`
COMPILER_PATH=$UE_ENGINE_DIR/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v15_clang-8.0.1-centos7/x86_64-unknown-linux-gnu
CMAKE=/opt/build/cmake/2.8.12.2/platform-linux/ext
OPENSSL=/opt/build/openssl/1.0.2k/platform-linux

export PATH=$PYTHON_INSTALL_DIR/bin:$COMPILER_PATH/bin:$CMAKE/bin:$OPENSSL/bin:$PATH
```
and set it on the top of the `build/BuildUsd.sh` file.

Good luck :)

**Comment by c64kernal** [comment]
Hi @sorenvdb -- I mentioned the Python issue because we've seen errors in the past where folks had multiple versions of Python and were building against one version and then using the runtime from another. I'm not sure if this is your problem, but I'd do what @msearl001 -- that seems much more likely to get you unblocked (thanks @msearl001!)

**Comment by sorenvdb** [comment]
Sorry, I forgot to mention I am not building for unreal, but just the USD library by itself. Also it was fixed by linking boost-python dynamically (which I mentioned in the issue I linked). Thanks for your input though.
